File: cli/gh-teacher/internal/configrepo/group_team_test.go
package configrepo

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/foundation50/classroom50-cli-shared/contract"
)

// TestMarshalGroupDescription_Golden pins the exact record bytes: the web
// writer mirrors Go's json.Marshal escaping, so a byte drift here silently
// diverges the two writers' descriptions.
func TestMarshalGroupDescription_Golden(t *testing.T) {
	got, err := MarshalGroupDescription("cs50", "project", "The Sharks")
	if err != nil {
		t.Fatalf("MarshalGroupDescription: %v", err)
	}
	want := `{"schema":"classroom50/group/v1","classroom":"cs50","assignment":"project","name":"The Sharks"}`
	if got != want {
		t.Errorf("record = %s, want %s", got, want)
	}

	// An empty display name is OMITTED, not written as "".
	got, err = MarshalGroupDescription("cs50", "project", "")
	if err != nil {
		t.Fatalf("MarshalGroupDescription: %v", err)
	}
	want = `{"schema":"classroom50/group/v1","classroom":"cs50","assignment":"project"}`
	if got != want {
		t.Errorf("nameless record = %s, want %s", got, want)
	}
}

// TestGroupDescription_SharedVectorParity runs the shared golden vectors: for
// every fixture case the marshalled record must parse back AND verify against
// the fixture's team-1 slug — tying the description record to the same
// hash derivation every writer pins.
func TestGroupDescription_SharedVectorParity(t *testing.T) {
	raw, err := os.ReadFile("../../../shared/testdata/group_vectors.json")
	if err != nil {
		t.Fatalf("read shared group vectors: %v", err)
	}
	var doc struct {
		Cases []struct {
			Classroom  string `json:"classroom"`
			Assignment string `json:"assignment"`
			Hash       string `json:"hash"`
			Team1      string `json:"team_1"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse shared group vectors: %v", err)
	}
	if len(doc.Cases) == 0 {
		t.Fatal("shared group vectors: no cases")
	}
	for _, c := range doc.Cases {
		desc, err := MarshalGroupDescription(c.Classroom, c.Assignment, "")
		if err != nil {
			t.Fatalf("MarshalGroupDescription(%q,%q): %v", c.Classroom, c.Assignment, err)
		}
		record, ok := ParseGroupDescription(desc)
		if !ok {
			t.Fatalf("ParseGroupDescription(%q) = !ok, want the record back", desc)
		}
		if !VerifyGroupDescription(c.Team1, record) {
			t.Errorf("VerifyGroupDescription(%q, record for %q/%q) = false, want true",
				c.Team1, c.Classroom, c.Assignment)
		}
		if got := contract.GroupTeamHash(c.Classroom, c.Assignment); got != c.Hash {
			t.Errorf("GroupTeamHash(%q,%q) = %q, want fixture %q", c.Classroom, c.Assignment, got, c.Hash)
		}
	}
}

// ParseGroupDescription is a trust boundary: absent/garbage/wrong-schema
// descriptions yield no record; unknown fields are tolerated (additive
// evolution).
func TestParseGroupDescription(t *testing.T) {
	valid := `{"schema":"classroom50/group/v1","classroom":"cs50","assignment":"project","name":"The Sharks"}`
	record, ok := ParseGroupDescription(valid)
	if !ok || record.Classroom != "cs50" || record.Assignment != "project" || record.Name != "The Sharks" {
		t.Errorf("ParseGroupDescription(valid) = (%+v, %t), want the full record", record, ok)
	}

	withUnknown := `{"schema":"classroom50/group/v1","classroom":"cs50","assignment":"project","future":"field"}`
	if _, ok := ParseGroupDescription(withUnknown); !ok {
		t.Error("a record with an unknown field must still parse (additive evolution)")
	}

	for _, bad := range []string{
		"",
		"   ",
		"ask the TA",
		`{"schema":"classroom50/invite/v1","classroom":"cs50","assignment":"project"}`,
		`{"schema":"classroom50/group/v1","classroom":"","assignment":"project"}`,
		`{"schema":"classroom50/group/v1","classroom":"cs50","assignment":""}`,
		`{"schema":"classroom50/group/v1"}`,
	} {
		if _, ok := ParseGroupDescription(bad); ok {
			t.Errorf("ParseGroupDescription(%q) = ok, want rejection", bad)
		}
	}
}

// VerifyGroupDescription must reject a record whose classroom/assignment
// doesn't hash back to the slug — the maintainer-edited-description defense —
// and any slug outside the full group-team shape.
func TestVerifyGroupDescription(t *testing.T) {
	slug := contract.GroupTeamName("cs50", "project", 3)
	good := GroupDescription{Schema: contract.GroupSchemaV1, Classroom: "cs50", Assignment: "project"}
	if !VerifyGroupDescription(slug, good) {
		t.Errorf("VerifyGroupDescription(%q, matching record) = false, want true", slug)
	}

	// A record pointing at a DIFFERENT assignment hashes elsewhere.
	stolen := GroupDescription{Schema: contract.GroupSchemaV1, Classroom: "cs50", Assignment: "other"}
	if VerifyGroupDescription(slug, stolen) {
		t.Error("a record for another assignment must not verify against this slug")
	}

	// Only the FULL slug shape is ever trusted for destructive ops.
	if VerifyGroupDescription("classroom50-group-theory", good) {
		t.Error("a human team in the classroom50-group- namespace must never verify")
	}
	if VerifyGroupDescription("classroom50-cs50", good) {
		t.Error("a classroom student team must never verify")
	}
}

File: web/src/util/groupTeam.test.ts
import { describe, expect, it } from "vitest"
import { readFileSync } from "node:fs"
import { fileURLToPath } from "node:url"
import {
  GROUP_HASH_HEX_LEN,
  GROUP_TEAM_PREFIX,
  groupTeamAssignmentPrefix,
  groupTeamHash,
  groupTeamName,
  isGroupTeamSlug,
  parseGroupTeamCounter,
} from "./teamSlug"
import {
  GROUP_DESCRIPTION_SCHEMA,
  marshalGroupDescription,
  parseGroupDescription,
  verifyGroupDescription,
} from "./groupTeam"
import { groupRepoName, parseGroupRepoCounter } from "./studentRepo"

// The shared vectors are the cross-language oracle (the Go suite asserts the
// same file), so a one-sided formula edit fails on the other language's copy.
const fixtureUrl = new URL(
  "../../../cli/shared/testdata/group_vectors.json",
  import.meta.url,
)
const doc = JSON.parse(readFileSync(fileURLToPath(fixtureUrl), "utf8")) as {
  hash_hex_len: number
  prefix: string
  cases: {
    classroom: string
    assignment: string
    hash: string
    team_1: string
  }[]
}

describe("group team naming (shared vectors)", () => {
  it("pins the fixture's contract halves", () => {
    expect(doc.hash_hex_len).toBe(GROUP_HASH_HEX_LEN)
    expect(doc.prefix).toBe(GROUP_TEAM_PREFIX)
    expect(doc.cases.length).toBeGreaterThan(0)
  })

  it.each(doc.cases)(
    "derives $team_1 for ($classroom, $assignment)",
    async ({ classroom, assignment, hash, team_1 }) => {
      expect(await groupTeamHash(classroom, assignment)).toBe(hash)
      expect(await groupTeamName(classroom, assignment, 1)).toBe(team_1)
      expect(isGroupTeamSlug(team_1)).toBe(true)
      const prefix = await groupTeamAssignmentPrefix(classroom, assignment)
      expect(parseGroupTeamCounter(team_1, prefix)).toBe(1)
    },
  )
})

describe("isGroupTeamSlug", () => {
  const hash = "cee5352cb880dff5"
  it.each([
    [`${GROUP_TEAM_PREFIX}${hash}-1`, true],
    [`${GROUP_TEAM_PREFIX}${hash}-42`, true],
    [`${GROUP_TEAM_PREFIX}${hash}-0`, false], // counters start at 1
    [`${GROUP_TEAM_PREFIX}${hash}-01`, false], // no leading zeros
    [`${GROUP_TEAM_PREFIX}${hash}-`, false],
    [`${GROUP_TEAM_PREFIX}${hash}`, false],
    [`${GROUP_TEAM_PREFIX}${hash.slice(0, 15)}-1`, false], // short hash
    ["classroom50-group-theory", false], // human team in the namespace
    ["classroom50-cs50", false], // classroom student team
    [`invite-${hash}`, false],
    [`x${GROUP_TEAM_PREFIX}${hash}-1`, false], // anchored
  ])("%s -> %s", (slug, want) => {
    expect(isGroupTeamSlug(slug)).toBe(want)
  })
})

describe("group repo naming", () => {
  it("composes and parses the counter, mode-gated by the caller", () => {
    expect(groupRepoName("CS101", "HW1", 3)).toBe("cs101-hw1-group-3")
    expect(parseGroupRepoCounter("cs101-hw1-group-3", "cs101", "hw1")).toBe(3)
    expect(parseGroupRepoCounter("CS101-HW1-GROUP-3", "cs101", "hw1")).toBe(3)
    for (const bad of [
      "cs101-hw1-group-0",
      "cs101-hw1-group-03",
      "cs101-hw1-group-",
      "cs101-hw1-group-x",
      "cs101-hw1-alice",
      "cs101-hw2-group-3", // other assignment
    ]) {
      expect(parseGroupRepoCounter(bad, "cs101", "hw1")).toBeNull()
    }
  })
})

describe("group description record", () => {
  it("round-trips and hash-verifies against the team slug", async () => {
    const meta = { classroom: "cs50", assignment: "project", name: "Sharks" }
    const record = parseGroupDescription(marshalGroupDescription(meta))
    expect(record).toEqual({ schema: GROUP_DESCRIPTION_SCHEMA, ...meta })
    const slug = await groupTeamName("cs50", "project", 2)
    expect(await verifyGroupDescription(slug, record!)).toBe(true)
    // A maintainer-edited record pointing at another assignment must fail the
    // re-hash check — the trust boundary for student-editable descriptions.
    const forged = parseGroupDescription(
      marshalGroupDescription({ classroom: "cs50", assignment: "other" }),
    )
    expect(await verifyGroupDescription(slug, forged!)).toBe(false)
  })

  it("omits an empty display name and tolerates junk descriptions", () => {
    expect(
      marshalGroupDescription({
        classroom: "cs50",
        assignment: "project",
        name: "  ",
      }),
    ).not.toContain('"name"')
    expect(parseGroupDescription(null)).toBeNull()
    expect(parseGroupDescription("not json")).toBeNull()
    expect(parseGroupDescription('{"schema":"classroom50/invite/v1"}')).toBe(
      null,
    )
  })
})
